mobzy-systems/src/main/kotlin/com/mineinabyss/mobzy/systems/listeners/LeashingListener.kt

```diff
@@ -0,0 +1,87 @@
+package com.mineinabyss.mobzy.systems.listeners
+
+import com.mineinabyss.geary.helpers.with
+import com.mineinabyss.geary.papermc.access.toGeary
+import com.mineinabyss.geary.papermc.access.toGearyOrNull
+import com.mineinabyss.mobzy.ecs.components.initialization.ModelEngineComponent
+import com.mineinabyss.mobzy.ecs.components.interaction.LeashEntity
+import com.mineinabyss.mobzy.systems.systems.ModelEngineSystem.toModelEntity
+import io.papermc.paper.event.entity.PufferFishStateChangeEvent
+import org.bukkit.entity.EntityType
+import org.bukkit.entity.LivingEntity
+import org.bukkit.entity.PufferFish
+import org.bukkit.event.EventHandler
+import org.bukkit.event.Listener
+import org.bukkit.event.entity.EntityDamageEvent
+import org.bukkit.event.entity.PlayerLeashEntityEvent
+import org.bukkit.event.player.PlayerUnleashEntityEvent
+import org.bukkit.potion.PotionEffectType
+
+object LeashingListener : Listener {
+
+    // Cancel Pufferfish change for leash-entities
+    @EventHandler
+    fun PufferFishStateChangeEvent.onLeash() {
+        isCancelled = entity.toGeary().has<LeashEntity>()
+    }
+
+    /** Handling leashing of entities with [ModelEngineComponent] */
+    @EventHandler
+    fun PlayerLeashEntityEvent.onLeashingMob() {
+        val gearyEntity = entity.toGearyOrNull() ?: return
+        val modelEntity = entity.toModelEntity() ?: return
+
+        // Leash is tied to the ModelEngine BaseEntity.
+        // This makes it exist for interaction but also invisible
+        // Spawn pufferfish, mount it to entity-bone and set leash-holder
+        gearyEntity.with { componentEntity: ModelEngineComponent ->
+            if (!componentEntity.leashable) return
+            if ((entity as? LivingEntity ?: return).isLeashed) return
+
+            val mount = modelEntity.mountHandler
+            val pufferFish = player.world.spawnEntity(entity.location, EntityType.PUFFERFISH) as PufferFish
+            pufferFish.puffState = 0
+            pufferFish.setAI(false)
+            pufferFish.isInvisible = true
+            pufferFish.isInvulnerable = true
+            pufferFish.toGeary().setPersisting(LeashEntity())
+
+            mount.setCanCarryPassenger(true)
```

Comment:
How come we set this here, i thought this was only for leashing?

Reply:
Well i add a component on it so that I can use said component to check if we should cancel the PufferFishState event.
Seemed like the cleanest way to handle it without nuking all pufferfishes no?

Reply:
I'm still not sure why ur specifically setting can carry, the component makes sense.